http-client/src/main/java/io/micronaut/http/client/MatchingFilterResolver.java

```diff
@@ -0,0 +1,171 @@
+import io.micronaut.core.annotation.AnnotationMetadataResolver;
+import io.micronaut.core.annotation.AnnotationValue;
+import io.micronaut.core.util.ArrayUtils;
+import io.micronaut.core.util.PathMatcher;
+import io.micronaut.core.util.StringUtils;
+import io.micronaut.core.util.Toggleable;
+import io.micronaut.http.HttpMethod;
+import io.micronaut.http.annotation.Filter;
+import io.micronaut.http.annotation.FilterAnnotation;
+import io.micronaut.http.filter.FilterProperties;
+import io.micronaut.http.filter.HttpClientFilter;
+import io.micronaut.http.filter.MatchingHttpClientFilter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.lang.annotation.Annotation;
+import java.net.URI;
+import java.util.*;
+import java.util.function.BooleanSupplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Resolver that responsible to finding all the http client filters that should be applied to given request.
+ *
+ * @since 1.2.4
+ * @author svishnyakoff
+ */
+@Primary
+@Singleton
+public class MatchingFilterResolver implements FilterResolver {
+    private static final Logger LOG = LoggerFactory.getLogger(MatchingFilterResolver.class);
+
+    private final Collection<HttpClientFilter> filters;
+    private final AnnotationMetadata annotationMetadata;
+    private final AnnotationMetadataResolver annotationMetadataResolver;
+    private final Set<String> clientIdentifiers;
+
+    /**
+     * @param annotationMetadata         annotation metadata of a declarative http client
+     * @param clientIdentifiers          client identifiers associated with a client
+     * @param annotationMetadataResolver annotation medata resolver
+     * @param filters                    all filters
+     */
+    @Inject
+    MatchingFilterResolver(@Nullable @Parameter AnnotationMetadata annotationMetadata,
+                           @Nullable @Parameter Set<String> clientIdentifiers,
+                           @Nullable AnnotationMetadataResolver annotationMetadataResolver,
+                           Collection<HttpClientFilter> filters) {
+        this.filters = filters;
+        this.annotationMetadata = Optional.ofNullable(annotationMetadata).orElse(AnnotationMetadata.EMPTY_METADATA);
+        this.annotationMetadataResolver = Optional.ofNullable(annotationMetadataResolver)
+                .orElse(AnnotationMetadataResolver.DEFAULT);
+        this.clientIdentifiers = Optional.ofNullable(clientIdentifiers).orElse(Collections.emptySet());
+    }
+
+    /**
+     * @param request    http request
+     * @param requestURI path URI
+     * @return filters that should be applied to request
+     */
+    public List<HttpClientFilter> resolveFilters(io.micronaut.http.HttpRequest<?> request,
+                                          URI requestURI) {
+        List<HttpClientFilter> filterList = new ArrayList<>();
```

Comment:
Pass a size to the list to save memory. It is unlikely that there will ever be more than 10 filters. So that seems reasonable number

Reply:
Will do